---
title: "Association with a sea anemone alters the skin microbiome of clownfish"
author: "Victoria French, Grace Beery & Madison Pacaro"
date: "March 9, 2021"
output: html_document
---


# Introduction

Sea anemone hosting temporarily alters clownfish skin microbiome. After analyzing the microbiome of clownfish not exposed to anemones, clownfish associated with anemones, and clownfish after being removed from their associated anemones, it was found that the microbiome of clownfish mucous is significantly different between hosting and non-hosting individuals. Interestingly, it was also found that a hosting clownfish's mucosal microbiome returns to that of a non-associated clownfish when removed from the associated anemone for a number of weeks (Pratte et al, 2018).  We analyzed a subset of the data presented by Pratte et al, to gain practice with RStudio metabarcoding and to test the replicability of the authors' genomic analysis. We used the DADA2 pipeline to filter and analyze Illumina raw sequence reads, to assign taxonomic classification via the SILVA v138 reference database, of anemone, clownfish and the associated water column. We utilized phyloseq to visualize this data and construct an ordinate plot based on Bray-Curtis dissimilarity to identify relatedness between sample microbiomes. We found that the microbial community was similar across analyses but lost specificity with the reduction of sample sizes in this replication. The differences found in taxonomic assignment contributed to inconclusive results when analyzing microbiome relatedness. Overall, this project was able to successfully replicate methods of the original analysis but was unable to produce the same results as were found in the original paper.

Pratte ZA, Patin NV, McWhirt ME, Caughman AM, Parris DJ, Stewart FJ (2018) Association with a sea anemone alters the skin microbiome of clownfish. Coral Reefs 37:1119-1125

# Version Control 

R version 4.0.3 was used for all analyses.

dada2 package version 1.18.0

ShortRead package version 1.48.0

ggplot2 package version 3.3.3

phyloseq package version 1.34.0

```{r, warning=FALSE, message=FALSE}
library(dada2); #packageVersion("dada2"); citation("dada2")
library(ShortRead); #packageVersion("ShortRead")
library(ggplot2); #packageVersion("ggplot2")
library(phyloseq); #packageVersion("phyloseq")
```

# Data Analysis & Methods

First we designated a path to the fastq files we will be analyzing.
```{r, results='hide'}
path <-"C:/Users/Maddy/Documents/BI586/fishmicrobes_2021_assign1/fastqfiles"
fns <- list.files(path)
fns
```

To divide fastq files into matched forward and reversed lists, determined by either a _1 or _2, we first had to ensure fastq files were in the same order. We then removed the .fastq from each file name.
```{r, results='hide'}
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs)
fnFs <- fastqs[grepl("_1", fastqs)]
fnRs <- fastqs[grepl("_2", fastqs)]

sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

```

We then set the full path to our forward and reverse reads.
```{r, results='hide'}
fnFs <- file.path(path, fnFs) 
fnRs <- file.path(path, fnRs)
fnFs
fnRs
```

Quality profiles were then plotted for forward and reverse reads to determine where to trim them.
```{r}
#Forward 1-9
plotQualityProfile(fnFs[1:9])
#Forward 10-18
plotQualityProfile(fnFs[10:18])
#Reverse 1-9
plotQualityProfile(fnRs[1:9])
#Reverse 10-18
plotQualityProfile(fnRs[10:18])
```

We then created a new subdirectory for trimmed and filtered files. 
```{r}
filt_path <- file.path(path, "trimmed")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

To filter the reads, we set trunclen to 240 because, despite some samples falling off extremely early (around cycle 100) no samples fell below a quality score of 20 after 250 bp.  

We determined that 240 maximizes the read output during inital filtering and minimizes presence of bimeras while still resulting in successful paired end merges.
We excluded the filtering parameter 'Trimleft' because our raw data did not contain primer sequences. 
The primer sequences used were

- 515F ('GTGYCAGCMGCCGCGGTAA')

- 806R ('GGACTACNVGGGTWTCTAAT')

for the 16S V4 region according to earth microbiome project.
```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=240,
                     maxN=0, 
                     maxEE=1, 
                     truncQ=2, 
                     rm.phix=TRUE,
                     compress=TRUE, multithread=FALSE) 
```
```{r}
head(out)
tail(out)
```

DADA2 was then utilized to employ a parametric error model to determine the error rate of the amplicon dataset. We kept the standard max consistency of 30 to increase the number of cycles and allow for convergence.
```{r}
setDadaOpt(MAX_CONSIST=30)
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
```

We plotted the found forward and reverse errors to visually confirm our data. The black line represents the estimated error rates, and the points represent the fit of the error rates. We should see a negative linear relationship in error rates, as quality score increases.
```{r, warning=FALSE, message=FALSE}
plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)
```

Reads were dereplicated, collapsing all reads into unique sequences. Then, the dereplicated reads were assigned to sample names.
```{r, results= 'hide'}
derepFs <- derepFastq(filtFs, verbose=FALSE)
derepRs <- derepFastq(filtRs, verbose=FALSE)

names(derepFs) <-sample.names
names(derepRs) <- sample.names
```

We used a band size of 16 to infer sequence variants as it is suggested as the default size for Illumina. 
```{r}
setDadaOpt(BAND_SIZE=16)
dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
```

To determine how many real variants are within unique input sequences, we looked at the dada class objects by sample.
Pooled processing for dada is available with the function pool=TRUE, which may provide better results for low sampling depths.
```{r}
dadaFs[[1]]
dadaRs[[1]]
```

Paired ends from forward and reverse reads were merged to obtain the full sequences. Denoised forward reads were aligned with reverse-compliment corresponding denoised reverse reads.
```{r, results='hide'}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=FALSE)
head(mergers[[1]])
```

We then constructed a sequence table. This table contains amplicon sequence variants (ASV).
```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

We then identified and removed chimeric sequences in our data.
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)
#chimeras account for only about ~1% of the merged sequence reads
```

We created a csv file to save the full sequences and sequences with chimeras removed.
```{r}
write.csv(seqtab,file="fishmicrobes_seqtab.csv")
write.csv(seqtab.nochim,file="fishmicrobes_nochim.csv")
```

**Tracking Read Stats**

Here, we are looking at the number of reads that made it through each step in the pipeline. We then created a csv table that displays the raw reads, how many reads were filtered, denoised, etc. The final number of reads is located under the nonchim column.

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaFs, getN), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track) <- sample.names 
head(track)
tail(track)

write.csv(track,file="ReadFilterStats_AllData_final.csv",row.names=TRUE,quote=FALSE)
```


# Assigning Taxonomy

We assigned taxonomy for amplicon sequence variants using the Silva v138 reference database. All parameters were conserved from the database presets.
```{r, results='hide'}
taxa <- assignTaxonomy(seqtab.nochim, "C:/Users/Maddy/Documents/BI586/fishmicrobes_2021_assign1/silva_nr99_v138_train_set.fa.gz", multithread=TRUE)
taxa <- addSpecies(taxa, "C:/Users/Maddy/Documents/BI586/fishmicrobes_2021_assign1/silva_species_assignment_v138.fa.gz")

taxa.print <- taxa
rownames(taxa.print) <- NULL #this gets rid of all the sequences
#head(taxa) #prints full info, including sequence
head(taxa.print) #this just prints taxonomy
```

We then created a csv file for the taxonomy so that it's easier to map the sequences for the heatmap.

```{r, results='hide'}
write.csv(taxa, file="taxa.csv",row.name=TRUE,quote=FALSE)
unname(head(taxa, 30))
unname(taxa)
```

Then, we saved our outputs so we can come back to the analysis stage at a later point if desired.
```{r, results='hide'}
saveRDS(seqtab.nochim, file="final_seqtab_nochim.rds")
saveRDS(taxa, file="final_taxa_blastCorrected.rds")
```

Here is how we can read in previously saved datafiles to avoid re-running the large datasets above.
```{r, results='hide'}
seqtab.nochim <- readRDS("final_seqtab_nochim.rds")
taxa <- readRDS("final_taxa_blastCorrected.rds")
#head(taxa)
```

We then imported our dataframe containing sample information. Before running this code, we made sure that sample names in the variable table had the same names as the sample names in the taxa and seqtab.nochim files. Samples should also be in the same order as the seqtab file in the rows, and variables as columns. 

```{r}
samdf<-read.csv("variabletable.csv")

head(samdf)
#head(otu_table(seqtab.nochim, FALSE))
#head(taxa)
rownames(samdf) <- samdf$Sample
```

We then constructed the phyloseq object.

```{r}
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE),
               sample_data(samdf),
               tax_table(taxa))
ps
```

We replaced sequences with names that correspond with the table output.
```{r}
ids<-taxa_names(ps)
ids <- paste0("sq",seq(1, length(colnames(seqtab.nochim))))
colnames(seqtab.nochim) <- ids
```

Then we created a Barplot to visualize sample microbe colonies.
```{r, warning=FALSE, message=FALSE}
top90 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:90]
#isolating top 90 sequences 
ps.top90 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
#transforming sequence counts to proportions for visualization
ps.top90 <- prune_taxa(top90, ps.top90)
#removing unwanted (not top90) OTUs (ASVs)
barplot<- plot_bar(ps.top90, x="Sample", fill="Class") 
```

To visualize counts rather than abundances:

```{r, warning=FALSE, message=FALSE}
barplot.count <- plot_bar(ps, x ="Sample", fill= "Class") #+ facet_wrap("tank")
```

We then created a csv file for the phyloseq data. This will give the abundances for each sample and class and is useful for constructing the heatmap. Also, it enables you to use ggplot, and construct more aesthetically pleasing plots.

```{r, warning=FALSE, message=FALSE}
psz <- psmelt(ps.top90)
write.csv(psz, file="Phyloseqoutputfinal.csv")
p <- ggplot(psz, aes(x=Treatment, y=Abundance, fill=Class))
p + geom_bar(stat="identity", colour="black")
```

Here we are looking at Bray Curtis dissimilarities to find if association with anemone has a significant difference on hosting vs. nonhosting clownfish microbiomes. 

```{r}
ord.nmds.bray <- ordinate(ps.top90, method="NMDS", distance="bray")
bray.curtis <-plot_ordination(ps.top90, ord.nmds.bray, color ='color', title="Bray NMDS")

```


# Figures

```{r}
p + geom_bar(stat="identity", colour="black")
```

Figure 1: Representation of proportional abundance of taxonomic groups within clownfish, anemone and water column microbiomes. Treatments categories are represented in the following format; IntialNumber. Where initial is indicating treatment (A-Anemone, H-Hosting Clownfish, NH-Non-hosting Clownfish, PH- Post-hosting Clownfish, WC-Water Column) and number indicates either week (in the case of clownfish) or individual (as for anemone and water column) samples. 

```{r}
bray.curtis
```

Figure 2: Relatedness of Anemone (A), Clownfish (NH,H,PH) and the water column (W) microbiomes according to NMDS analysis based on Bray-Curtis dissimilarities. 

# Reference Database

Reference database used was Silva SSU taxonomic training data formatted for DADA2 (Silva version 138) as sequences were assessed by Illumina sequencing of the 16S rRNA gene (V4 region). Original analysis done on silva v119. 

Citation: 

@dataset{michael_r_mclaren_2020_3986799,
  author       = {Michael R. McLaren},
  title        = {{Silva SSU taxonomic training data formatted for 
                   DADA2 (Silva version 138)}},
  month        = aug,
  year         = 2020,
  note         = {{These files are intended for use in classifying 
                   prokaryotic 16S sequencing data and are not
                   appropriate for classifying eukaryotic ASVs. The
                   new method implemented within DADA2 for
                   constructing these files only includes 100
                   eukaryotic sequences for use as an outgroup.}},
  publisher    = {Zenodo},
  version      = 2,
  doi          = {10.5281/zenodo.3986799},
  url          = {https://doi.org/10.5281/zenodo.3986799}


# Conclusion

Overall, our project was unable to conclusively produce the same results as originally presented in Pratte et al. We found similar taxonomic classification, yet our assignments were of broader scale and the presence of specific classes were lost entirely. This could be due to a multiple of factors related to sampling. We used a much smaller sampling size across the three sampling types (A.clarkii n=3, water column n=2, anemone n=2) when running our analyses compared to the original paper (A.clarkii n=12, water column n=9, anemone n=22). Under the assumption that every individual's microbiome is differentiated due to conditional inputs and random chance, it is likely that the response of the hosting treatment individual is not indicative of a larger population response. There is also a chance that the sequenced anemone and the sequenced hosting clownfish had zero interaction history as association between specific colonies and A.clarkii was not recorded in metadata. This could account for the differences we see in taxonomic makeup between our anemone and hosting clownfish microbiome communities. The higher order classifications led to vastly different Bray-Curtis relationships in our data than in the metadata. There are very weak clustering patterns overall with non-hosting outliers indicating the greatest differences. While the some non-hosting, hosting and post-hosting individuals maintain relative similarity to the anemone samples. Therefore we can not conclusively say that association with anemone alters the skin microbiome of a clownfish. 


